data-prepper-plugins/kafka-source/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/consumer/PlainTextConsumer.java

```diff
@@ -0,0 +1,89 @@
+
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.kafka.clients.consumer.CommitFailedException;
+import org.apache.kafka.clients.consumer.ConsumerRebalanceListener;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.common.TopicPartition;
+import org.opensearch.dataprepper.metrics.PluginMetrics;
+import org.opensearch.dataprepper.model.buffer.Buffer;
+import org.opensearch.dataprepper.model.record.Record;
+import org.opensearch.dataprepper.plugins.kafka.source.KafkaSourceBuffer;
+import org.opensearch.dataprepper.plugins.kafka.source.configuration.TopicConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * * A helper class which helps to process the Plain Text records and write them
+ * into the buffer. Offset handling and consumer re balance are also being
+ * handled
+ */
+
+@SuppressWarnings("deprecation")
+public class PlainTextConsumer implements KafkaSchemaTypeConsumer<String, String>, ConsumerRebalanceListener {
+  private static final Logger LOG = LoggerFactory.getLogger(PlainTextConsumer.class);
+  private Map<TopicPartition, OffsetAndMetadata> currentOffsets = new HashMap<>();
+  private long lastReadOffset = 0L;
+  private KafkaConsumer<String, String> plainTxtConsumer;
+
+  @SuppressWarnings({ "unchecked", "rawtypes" })
+  @Override
+  public void consumeRecords(KafkaConsumer<String, String> consumer, AtomicBoolean status,
+      Buffer<Record<Object>> buffer, TopicConfig topicConfig, PluginMetrics pluginMetrics) {
+    KafkaSourceBuffer kafkaSourceBuffer = new KafkaSourceBuffer(topicConfig, pluginMetrics);
+    plainTxtConsumer = consumer;
+    try {
+      currentOffsets.clear();
+      consumer.subscribe(Arrays.asList(topicConfig.getName()));
+      while (!status.get()) {
+        ConsumerRecords<String, String> records = consumer.poll(Duration.ofMillis(100));
+        if (!records.isEmpty() && records.count() > 0) {
+          for (TopicPartition partition : records.partitions()) {
+            List<ConsumerRecord<String, String>> partitionRecords = records.records(partition);
+            for (ConsumerRecord<String, String> consumerRecord : partitionRecords) {
+              currentOffsets.put(new TopicPartition(consumerRecord.topic(), consumerRecord.partition()),
+                 new OffsetAndMetadata(consumerRecord.offset() + 1, null));
+              kafkaSourceBuffer.publishRecordToBuffer(consumerRecord.value(), buffer, topicConfig);
+              lastReadOffset = partitionRecords.get(partitionRecords.size() - 1).offset();
+              kafkaSourceBuffer.commitOffsets(partition, lastReadOffset, consumer);
```

Comment:
Offsets commits on every read is expensive.  I would have a periodic commits of offsets to improve efficiency. 

Define commitOffsets() method and call it periodically.  Refer good article :  https://www.confluent.io/blog/kafka-consumer-multi-threaded-messaging/

Reply:
Resolved.
As of now we have a periodic offset commits every 5 minutes.